segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentStateStore.java

```diff
@@ -72,36 +91,109 @@
 
     @Override
     public CompletableFuture<SegmentState> get(String segmentName, Duration timeout) {
-        String stateSegment = StreamSegmentNameUtils.getStateSegmentName(segmentName);
-        TimeoutTimer timer = new TimeoutTimer(timeout);
-        return this.storage
-                .getStreamSegmentInfo(stateSegment, timer.getRemaining())
-                .thenComposeAsync(sp -> readSegmentState(sp, timer.getRemaining()), this.executor)
-                .exceptionally(this::handleSegmentNotExistsException);
+        return findLatestValidState(segmentName, timeout)
+                .thenApplyAsync( pair -> {
+                    return  pair.getRight();
+                });
+    }
+
+    private CompletableFuture<Pair<SegmentProperties, SegmentState>> findLatestValidState(String segmentName, Duration timeout) {
+        String stateSegment1 = StreamSegmentNameUtils.getFirstStateSegmentName(segmentName);
+        String stateSegment2 = StreamSegmentNameUtils.getSecondStateSegmentName(segmentName);
+
+        return getPropertiesAndState(stateSegment1, timeout)
+                .thenCombineAsync(getPropertiesAndState(stateSegment2, timeout),
+                        (first, second) -> {
+                            SegmentProperties property1 = first.getLeft();
+                            SegmentState state1 = first.getRight();
+                            SegmentProperties property2 = second.getLeft();
+                            SegmentState state2 = second.getRight();
+                            if (property1 == null) {
+                                return Pair.of(property2, state2);
+                            }
+                            if (property2 == null) {
+                                return Pair.of(property1, state1);
+                            }
+                            if (property1.getLastModified().asDate().compareTo(property2.getLastModified().asDate()) > 0) {
+                                return Pair.of(property1, state1);
+                            } else {
+                                return Pair.of(property2, state2);
+                            }
+                        }, this.executor);
+    }
+
+    private CompletableFuture<Pair<SegmentProperties, SegmentState>> getPropertiesAndState(String stateSegment, Duration timeout) {
+        final SegmentProperties[] segmentProperties = new SegmentProperties[1];
```

Comment:
Why an array of size 1?

Reply:
I think this would be better served with an AtomicReference. The semantics of that are much clearer